File: LeetCode/20-valid-parentheses/ex_020.java
public class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            switch (c) {
                case '(':
                case '[':
                case '{':   stack.push(c);  break;
                case ')':   if (stack.empty() || stack.peek() != '(')   return false; else stack.pop(); break;
                case ']':   if (stack.empty() || stack.peek() != '[')   return false; else stack.pop(); break;
                case '}':   if (stack.empty() || stack.peek() != '{')   return false; else stack.pop(); break;
                default:    break;
            }
        }
        return stack.empty();
    }
}

File: LeetCode/20-valid-parentheses/ex_020.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        pairs = [('(', ')'), ('[', ']'), ('{', '}')]
        for c in s:
            if c in '([{':
                stack.append(c)
            elif not stack:
                return False
            else:
                for l, r in pairs:
                    if c == r and stack.pop() != l:
                        return False
        return not stack